octodns/provider/ns1.py

```diff
@@ -27,11 +27,41 @@ class Ns1Exception(Exception):
 class Ns1Client(object):
     log = getLogger('NS1Client')
 
-    def __init__(self, api_key, retry_count=4):
-        self.log.debug('__init__: retry_count=%d', retry_count)
+    def __init__(self, api_key, parallelism=None, retry_count=4):
+        self.log.debug('__init__: parallelism=%s, retry_count=%d', parallelism,
+                       retry_count)
         self.retry_count = retry_count
 
         client = NS1(apiKey=api_key)
+
+        # NS1 rate limits via a "token bucket" scheme, and provides information
+        # about rate limiting in headers on responses. Token bucket can be
+        # thought of as an initially "full" bucket, where, if not full, tokens
+        # are added at some rate. This allows "bursting" requests until the
+        # bucket is empty, after which, you are limited to the rate of token
+        # replenishment.
+        # There are a couple of "strategies" built into the SDK to avoid 429s
+        # from rate limiting. Since octodns operates concurrently via
+        # `max_workers`, a concurrent strategy seems appropriate.
+        # This strategy does nothing until the remaining requests are equal to
+        # or less than our `parallelism`, after which, each process will sleep
+        # for the token replenishment interval times parallelism.
+        # For example, if we can make 10 requests in 60 seconds, a token is
+        # replenished every 6 seconds. If parallelism is 3, we will burst 7
+        # requests, and subsequently each process will sleep for 18 seconds
+        # before making another request.
+        # In general, parallelism should match the number of workers.
+        if parallelism is not None:
+            client.config['rate_limit_strategy'] = 'concurrent'
+            client.config['parallelism'] = parallelism
```

Comment:
Would it ever make sense to use another strategy (with I'd assume other prams) with octoDNS or would it always be `concurrent`?

Reply:
there's only one other one in the SDK right now, and I just don't see it ever being better than this one for octodns. As far as i can tell, assuming that we don't get rate limited in the `plan` stage, a `parallelism` of 1 would be as good as it gets for saturating the connection.
for now, I wanted to keep it simple, and i have a hope that as we make improvements this strategy will still be the one that makes the most sense, and they'll come for free with version bumps.

do you have any recommendations on making these sort of provider caveats more visible? We'd be happy to take some responsibility for maintaining ours, as would others, I'm sure.  I thought maybe if you stuck each provider in a their own dir, there could be a README in each one, and github would serve it really nicely :)

Reply:
It was more that if there are other useful options (or likely to be others in the future) it might make sense to expose a lower-level option, like a `client_config={}` param. If there's not anything useful now we can leave that for if/when.

> do you have any recommendations on making these sort of provider caveats more visible? We'd be happy to take some responsibility for maintaining ours, as would others, I'm sure. I thought maybe if you stuck each provider in a their own dir, there could be a README in each one, and github would serve it really nicely :)

The actual plan, though no clue when I'll get a chance to make it happen, is to break up each of the providers into their own repo under an org https://github.com/octodns. That way they can have their own contributors, releases, ...

Reply:
Ah yeah, great point. there are indeed a couple of client config options (not explicitily related to rate-limiting), that I'd like to add support for. I was going to make a lil PR for them once this settled, but having a `client_config={}` param would take care of that in a flexible way.
 
We could also handle the current rate limiting approach this way, but since the settings for it are a bit fiddly, I'd prefer to leave the current `parallelism` setting in place. 

I'll have a patch up shortly, I think I can make it pretty tidy